### PR TITLE
Improve README

### PR DESCRIPTION
**Pull request by ajnavarro**
Signed-off-by: Antonio Jesus Navarro Perez <antnavper@gmail.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,11 +41,17 @@ We are continuously adding more functionality to go-mysql-server. We support a s
...
+- `SUBSTRING(str,pos)`, ` SUBSTRING(str,pos,len)`: return a substring from the provided string.
+- Date and Timestamp functions: `YEAR(date)`, `MONTH(date)`, `DAY(date)`, `HOUR(date)`, `MINUTE(date)`, `SECOND(date)`, `DAYOFYEAR(date)`.
+- `ARRAY_LENGTH(json)`: If the json representation is an array, this function returns its size.
 
 ## Example
 
-`go-mysql-server` has a sql engine and a server implementation, so to start a server you must instantiate the engine and give it your `sql.Database` implementation that will be in charge to handle all the logic about retrieving the data from your source :
+`go-mysql-server` has a sql engine and a server implementation,
+so to start a server you must instantiate the engine and give it your `sql.Database` implementation.
+It will be in charge to handle all the logic about retrieving the data from your source.
```
Comment (erizocosmico) [inline README.md]: s/to handle/of handling/

Comment (erizocosmico) [reply README.md]: s/about retrieving/for retrieving/

File: README.md [hunk 2]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
 | evilbob@gmail.com |
 +-------------------+
 ```
+## Custom data source implementation
+
+To be able to create your own data source implementation you need to implement the following interfaces:
+
+- `sql.Database` interface. This interface will provide tables from your data source.
+  - If your database implementation supports add more tables, you might want to add support to `sql.Alterable` interface
```
Comment (erizocosmico) [inline README.md]: s/supports add more tables/supports adding more tables/
s/add support to/add support for/

File: README.md [hunk 3]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
...
+
+- `sql.Table` interface. It will be in charge to transform any kind of data to an iterator of Rows. Depending of how much you want to optimize the queries, you also can implement other interfaces on your tables:
```
Comment (erizocosmico) [inline README.md]: s/to transform/of transforming/
s/to an iterator/into an iterator/

File: README.md [hunk 4]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
...
+  - `sql.PushdownProjectionTable` interface will provides you a way to get only the columns needed for the executed query.
```
Comment (erizocosmico) [inline README.md]: s/will provides/will provide/

Comment (erizocosmico) [reply README.md]: s/you a way/a way/

File: README.md [hunk 5]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
...
+  - `sql.PushdownProjectionAndFiltersTable` interface will provide the same functionality described before, but also will push down the filters used in the executed query. Doing this, it allows you to filter before hand data, and improve query speed.
```
Comment (erizocosmico) [inline README.md]: s/it allows you to/you can/
s/before hand data/data beforehand/
s/query speed/the query speed/

File: README.md [hunk 6]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
...
+  - `sql.Indexable` add index capabilities to your table. Implementing this interface you can create and use indexes on this tables.
```
Comment (erizocosmico) [inline README.md]: s/Implementing this/By implementing this/
s/on this tables/on this table/

File: README.md [hunk 7]
```diff
@@ -136,11 +146,26 @@ SELECT email FROM mytable WHERE name = 'Evil Bob'
...
+  - `sql.Inserter` can be implemented if your data source tables allows insertions.
```
Comment (erizocosmico) [inline README.md]: s/allows/allow/

File: README.md [hunk 8]
```diff
@@ -37,11 +41,17 @@ We are continuously adding more functionality to go-mysql-server. We support a s
...
-- `IS_BINARY(blob)`: returns whether a BLOB is a binary file or not
+- `IS_BINARY(blob)`: returns whether a BLOB is a binary file or not.
+- `SUBSTRING(str,pos)`, ` SUBSTRING(str,pos,len)`: return a substring from the provided string.
+- Date and Timestamp functions: `YEAR(date)`, `MONTH(date)`, `DAY(date)`, `HOUR(date)`, `MINUTE(date)`, `SECOND(date)`, `DAYOFYEAR(date)`.
+- `ARRAY_LENGTH(json)`: If the json representation is an array, this function returns its size.
 
 ## Example
 
-`go-mysql-server` has a sql engine and a server implementation, so to start a server you must instantiate the engine and give it your `sql.Database` implementation that will be in charge to handle all the logic about retrieving the data from your source :
+`go-mysql-server` has a sql engine and a server implementation,
```
Comment (kuba--) [inline README.md]: I would write sth. like:
>>
...contains a sql engine and server implementation. So, if you want to start a server, first instantiate the engine and pass your `sql.Database` implementation.

File: README.md (hunk 1 above)
Comment (kuba--) [reply README.md]: or maybe "...to retrieve the data from your source"

File: README.md (hunk 5 above)
Comment (kuba--) [reply README.md]: Maybe:
>> 
It allows to filter a data in advance, and speed up query
>>

**Comment by ajnavarro** [comment]
@erizocosmico done.

**Pull request merged**